wagtail/admin/views/pages/edit.py

```diff
@@ -375,79 +379,38 @@ def dispatch(self, request, page_id):
-            # this should be indicated separately
-            if self.page.locked_by and self.page.locked_at:
-                lock_message = format_html(
-                    _("<b>Page '{}' was locked</b> by <b>{}</b> on <b>{}</b>."),
-                    self.page.get_admin_display_title(),
-                    str(self.page.locked_by),
-                    self.page.locked_at.strftime("%d %b %Y %H:%M"),
-                )
-            else:
-                # Page was probably locked with an old version of Wagtail, or a script
-                lock_message = format_html(
-                    _("<b>Page '{}' is locked</b>."),
-                    self.page.get_admin_display_title(),
-                )
+        if self.lock:
+            lock_message = self.lock.get_message(self.request.user)
+            if lock_message:
+                if isinstance(self.lock, BasicLock) and self.page_perms.can_unlock():
+                    lock_message = format_html(
+                        '{} <span class="buttons"><button type="button" class="button button-small button-secondary" data-action-lock-unlock data-url="{}">{}</button></span>',
+                        lock_message,
+                        reverse("wagtailadmin_pages:unlock", args=(self.page.id,)),
+                        _("Unlock"),
+                    )
 
-            if self.page_perms.can_unlock():
-                lock_message += format_html(
-                    '<span class="buttons"><button type="button" class="button button-small button-secondary" data-action-lock-unlock data-url="{}">{}</button></span>',
-                    reverse("wagtailadmin_pages:unlock", args=(self.page.id,)),
-                    _("Unlock"),
-                )
-            messages.error(self.request, lock_message, extra_tags="lock")
-
-        if self.page.current_workflow_state:
-            workflow = self.workflow_state.workflow
-            task = self.workflow_state.current_task_state.task
-            if (
-                self.workflow_state.status != WorkflowState.STATUS_NEEDS_CHANGES
-                and task.specific.page_locked_for_user(self.page, self.request.user)
-            ):
-                # Check for revisions still undergoing moderation and warn
-                if len(self.workflow_tasks) == 1:
-                    # If only one task in workflow, show simple message
-                    workflow_info = _("This page is currently awaiting moderation.")
-                else:
-                    workflow_info = format_html(
-                        _(
-                            "This page is awaiting <b>'{}'</b> in the <b>'{}'</b> workflow."
+                if (
+                    isinstance(self.lock, ScheduledForPublishLock)
+                    and self.page_perms.can_unschedule()
+                ):
+                    scheduled_revision = self.page.revisions.defer().get(
+                        approved_go_live_at__isnull=False
+                    )
```

Comment:
FYI I was able to reduce a considerable amount of duplicated queries (around 5) by turning this into a cached property in `DraftStateMixin`, and using that here and in `get_message()`, which I think should be safe in these cases.

See https://github.com/wagtail/wagtail/pull/8901/commits/01711e631df64a23a82e5364786420bb2efa3914